### PR TITLE
SMHE-1844-Get-and-SetConfigurationObjectCommandExecutor

### PR DESCRIPTION
**Pull request by harrymiddelburgetall**
- ObjectConfigService implemented for:

  - GetConfigurationObjectCommandExecutor
  - SetConfigurationObjectCommandExecutor
  - SetRandomisationSettingsCommandExecutor

- Added cucumber tests for all (D)SMR protocol versions

Bycatch:

- Removed inconsistant usage of PO (with capital o) and P0 (with zero) for ConfigurationObjectFlagType in xsd

```
  <!-- ConfigurationObjectFlagType -->
  <xsd:simpleType name="ConfigurationFlagType">
    <xsd:restriction base="xsd:string">

      <xsd:enumeration value="HLS_3_ON_PO_ENABLE" />
      <xsd:enumeration value="HLS_4_ON_PO_ENABLE" />
      <xsd:enumeration value="HLS_5_ON_PO_ENABLE" />

      <xsd:enumeration value="HLS_6_ON_P0_ENABLE" />
      <xsd:enumeration value="HLS_7_ON_P0_ENABLE" />
```
	  

- Add support for DIRECT_ATTACH_AT_POWER_ON in protocol version DSMR 4.3

### CODE REVIEW

**Comment by harrymiddelburgetall** [comment]
✔ https://jenkins.fdp.osgp.cloud/view/SMHE-cucumber/job/gxf-smhe-cucumber-build/173/

File: osgp/protocol-adapter-dlms/osgp-dlms/src/main/resources/dlmsprofiles/dlmsprofile-dsmr422.json [hunk 1]
```diff
@@ -2036,6 +2036,46 @@
...
+      "properties": {},
+      "attributes": [
+        {
+          "id": 2,
+          "description": "value",
+          "datatype": "structure",
+          "valuetype": "SET_BY_CLIENT",
+          "access": "RW"
+        }
+      ]
```
Comment (stefanermens) [inline osgp/protocol-adapter-dlms/osgp-dlms/src/main/resources/dlmsprofiles/dlmsprofile-dsmr422.json]: This object is not specified in DSMR4.2.2, but is needed. Please add a note explaining that this object should really be in this config.

File: integration-tests/cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/configuration/SetConfigurationObject.feature [hunk 2]
```diff
@@ -72,17 +98,21 @@ Feature: SmartMetering Configuration
       | ConfigurationFlagEnabled_9  | false                  |
       | ConfigurationFlagType_10    | HLS_5_ON_PO_ENABLE     |
       | ConfigurationFlagEnabled_10 | false                  |
+    Examples:
+      | deviceIdentification |
+      | TEST1024000000001    |
+      | TEST1027000000001    |
```
Comment (stefanermens) [inline integration-tests/cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/configuration/SetConfigurationObject.feature]: Should we test other protocol versions as well? For example, what should a DSMR2.2 meter do?

File: osgp/protocol-adapter-dlms/osgp-dlms/src/main/resources/dlmsprofiles/dlmsprofile-dsmr422.json (hunk 1 above)
Comment (harrymiddelburgetall) [reply osgp/protocol-adapter-dlms/osgp-dlms/src/main/resources/dlmsprofiles/dlmsprofile-dsmr422.json]: Added this note:
_This object configuration is supported by DSMR4 devices but not presented in de DSMR4 specification. This object configuration is added according to the SMR 5.0.0 specification_

**Comment by harrymiddelburgetall** [comment]
✔ https://jenkins.fdp.osgp.cloud/view/SMHE-cucumber/job/gxf-smhe-cucumber-build/184/

File: integration-tests/cucumber-tests-platform-smartmetering/src/test/java/org/opensmartgridplatform/cucumber/platform/smartmetering/glue/steps/ws/smartmetering/smartmeteringbundle/BundledGetConfigurationObjectSteps.java [hunk 3]
```diff
@@ -49,9 +50,14 @@ public void theBundleResponseShouldContainAConfigurationObjectResponse(
     final ConfigurationObject configurationObject =
         ((GetConfigurationObjectResponse) response).getConfigurationObject();
 
-    assertThat(configurationObject.getGprsOperationMode().toString())
-        .as("The gprs operation mode is not equal")
-        .isEqualTo(values.get("GprsOperationMode"));
+    if (values.containsKey("GprsOperationMode")
```
Comment (kroesctrl) [inline integration-tests/cucumber-tests-platform-smartmetering/src/test/java/org/opensmartgridplatform/cucumber/platform/smartmetering/glue/steps/ws/smartmetering/smartmeteringbundle/BundledGetConfigurationObjectSteps.java]: Make static variable "GprsOperationMode"

File: integration-tests/cucumber-tests-platform-smartmetering/src/test/java/org/opensmartgridplatform/cucumber/platform/smartmetering/support/ws/smartmetering/configuration/ConfigurationObjectFactory.java [hunk 4]
```diff
@@ -4,6 +4,7 @@
 
 package org.opensmartgridplatform.cucumber.platform.smartmetering.support.ws.smartmetering.configuration;
 
+import io.micrometer.common.util.StringUtils;
```
Comment (kroesctrl) [inline integration-tests/cucumber-tests-platform-smartmetering/src/test/java/org/opensmartgridplatform/cucumber/platform/smartmetering/support/ws/smartmetering/configuration/ConfigurationObjectFactory.java]: Use a othernmore generic StringUtils

File: integration-tests/cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/configuration/SetRandomisationSettings.feature [hunk 5]
```diff
@@ -7,47 +7,24 @@ Feature: SmartMetering Configuration - SetRandomisationSettings
...
+      | DeviceIdentification     | <deviceIdentification> |
+      | directAttach             | <da>                   |
+      | randomisationStartWindow | <rsw>                  |
+      | multiplicationFactor     | <mf>                   |
+      | numberOfRetries          | <nor>                  |
+    Then the randomisation settings <shouldBeSetOrNot> on the device
+    Examples:
+      | deviceIdentification | protocol | version | da | rsw | mf | nor | shouldBeSetOrNot  |
+      | TEST1027000000001    | SMR      | 5.1     |  1 |   1 |  1 |   1 | should be set     |
+      | TEST1027000000001    | SMR      | 5.1     |  0 |   5 |  3 |   1 | should be set     |
```
Comment (kroesctrl) [inline integration-tests/cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/configuration/SetRandomisationSettings.feature]: Please add SMR 5.0.0

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/configuration/service/GetConfigurationObjectServiceDsmr4.java [hunk 6]
```diff
@@ -33,13 +36,17 @@ public class GetConfigurationObjectServiceDsmr4 extends GetConfigurationObjectSe
...
+  public GetConfigurationObjectServiceDsmr4(
+      final DlmsHelper dlmsHelper, final ObjectConfigServiceHelper objectConfigServiceHelper) {
     this.dlmsHelper = dlmsHelper;
+    this.objectConfigServiceHelper = objectConfigServiceHelper;
   }
 
   @Override
   public boolean handles(final Protocol protocol) {
-    return protocol == Protocol.DSMR_4_2_2;
+    return protocol != null && protocol.isDsmr4() && !"4.3".equals(protocol.getVersion());
```
Comment (kroesctrl) [inline osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/configuration/service/GetConfigurationObjectServiceDsmr4.java:49]: "4.2".equals

**Review by kroesctrl — CHANGES_REQUESTED** [review]

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/configuration/service/SetConfigurationObjectServiceDsmr43.java [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+public class SetConfigurationObjectServiceDsmr43 extends SetConfigurationObjectServiceDsmr4 {
+
+  public SetConfigurationObjectServiceDsmr43(
+      final DlmsHelper dlmsHelper, final ObjectConfigServiceHelper objectConfigServiceHelper) {
+    super(dlmsHelper, objectConfigServiceHelper);
+  }
+
+  @Override
+  public boolean handles(final Protocol protocol) {
+    return protocol != null && protocol.isDsmr4() && "4.3".equals(protocol.getVersion());
```
Comment (stefanermens) [inline osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/configuration/service/SetConfigurationObjectServiceDsmr43.java]: Isn't `protocol == Protocol.DSMR_4_3` more simple?

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/test/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/configuration/GetConfigurationObjectCommandExecutorTest.java [hunk 8]
```diff
@@ -22,26 +23,29 @@
...
   @Mock private DlmsConnectionManager conn;
   @Mock private ConfigurationObjectDto configurationObjectDto;
   @Mock private GetConfigurationObjectService getService;
 
-  @Test
-  public void execute() throws ProtocolAdapterException {
+  @ParameterizedTest
+  @EnumSource(
+      value = Protocol.class,
+      names = {"DSMR_4_2_2", "SMR_5_0_0"})
```
Comment (stefanermens) [inline osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/test/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/configuration/GetConfigurationObjectCommandExecutorTest.java]: Should DSMR4.3 be tested as well?

**Review by stefanermens — CHANGES_REQUESTED** [review]

File: integration-tests/cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/configuration/SetConfigurationObject.feature (hunk 2 above)
Comment (harrymiddelburgetall) [reply integration-tests/cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/configuration/SetConfigurationObject.feature]: Added cucumber tests for all (D)SMR protocol versions

File: integration-tests/cucumber-tests-platform-smartmetering/src/test/java/org/opensmartgridplatform/cucumber/platform/smartmetering/glue/steps/ws/smartmetering/smartmeteringbundle/BundledGetConfigurationObjectSteps.java (hunk 3 above)
Comment (harrymiddelburgetall) [reply integration-tests/cucumber-tests-platform-smartmetering/src/test/java/org/opensmartgridplatform/cucumber/platform/smartmetering/glue/steps/ws/smartmetering/smartmeteringbundle/BundledGetConfigurationObjectSteps.java]: static final String GPRS_OPERATION_MODE = "GprsOperationMode";

File: integration-tests/cucumber-tests-platform-smartmetering/src/test/java/org/opensmartgridplatform/cucumber/platform/smartmetering/support/ws/smartmetering/configuration/ConfigurationObjectFactory.java (hunk 4 above)
Comment (harrymiddelburgetall) [reply integration-tests/cucumber-tests-platform-smartmetering/src/test/java/org/opensmartgridplatform/cucumber/platform/smartmetering/support/ws/smartmetering/configuration/ConfigurationObjectFactory.java]: import org.apache.commons.lang3.StringUtils;

File: integration-tests/cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/configuration/SetRandomisationSettings.feature (hunk 5 above)
Comment (harrymiddelburgetall) [reply integration-tests/cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/configuration/SetRandomisationSettings.feature]: Now testing 
| deviceIdentification | protocol | version |
 TEST1028000000001    | SMR      | 5.1     |
| TEST1027000000001    | SMR      | 5.0.0   |

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/configuration/service/SetConfigurationObjectServiceDsmr43.java (hunk 7 above)
Comment (harrymiddelburgetall) [reply osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/configuration/service/SetConfigurationObjectServiceDsmr43.java]: implemented isDsmr42() and isDsmr43() on Protocol class

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/test/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/configuration/GetConfigurationObjectCommandExecutorTest.java (hunk 8 above)
Comment (harrymiddelburgetall) [reply osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/test/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/configuration/GetConfigurationObjectCommandExecutorTest.java]: removed the names attribute from @EnumSource. All protocols are tested

File: integration-tests/cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/configuration/SetConfigurationObject.feature (hunk 2 above)
Comment (harrymiddelburgetall) [reply integration-tests/cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/configuration/SetConfigurationObject.feature]: DSMR 2.2 is now tested to fail

File: integration-tests/cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/configuration/SetRandomisationSettings.feature [hunk 9]
```diff
@@ -7,47 +7,26 @@ Feature: SmartMetering Configuration - SetRandomisationSettings
   As a grid operator
   I want to be able to set randomisation settings on a device
 
-  Scenario: Set direct attach on a SMR5 device
+  Scenario Outline: Set set randomisation settings on a <protocol> <version> device with direct attach <da>
```
Comment (kroesctrl) [inline integration-tests/cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/configuration/SetRandomisationSettings.feature]: Set set

**Review by kroesctrl — CHANGES_REQUESTED** [review]

**Review by stefanermens — APPROVED** [review]

File: integration-tests/cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/configuration/SetRandomisationSettings.feature (hunk 9 above)
Comment (harrymiddelburgetall) [reply integration-tests/cucumber-tests-platform-smartmetering/src/test/resources/features/osgp-adapter-ws-smartmetering/configuration/SetRandomisationSettings.feature]: Scenario Outline: Set randomisation settings on a <protocol> <version> device with direct attach <da>

**Review by kroesctrl — APPROVED** [review]

**Pull request merged**